### PR TITLE
Implement AsyncInterop\Promise

### LINKED ISSUE #64: Interoperability Implementation
Do you plan to implement https://github.com/async-interop/awaitable? Would you mind to provide feedback?

### PR DESCRIPTION
**Pull request by jsor**
Interoperability with [async-interop/promise](https://github.com/async-interop/promise).

Closes #64

### CODE REVIEW

**Comment by trowski** [comment]
@jsor I fixed an issue with the `@requires` annotation in `async-interop/promise-test`. If you restart travis the tests should all pass this time.

**Comment by jsor** [comment]
@trowski Thanks 👍

**Comment by WyriHaximus** [comment]
FYI: this effects this issue https://github.com/async-interop/event-loop/issues/128

**Comment by jsor** [comment]
@WyriHaximus Upgraded to async-interop/promise ^0.4.0 with the new namespace.

**Comment by jsor** [comment]
@clue I don't see this as an issue. I intentionally left out any documention of `when()` because i don't consider it part of the API of react/promise (while writing this, i realized 4b42c60 was wrong and reverted it). A consumer should not use `when()`, they should `then()` because that is our API.

`when()` is only for interoperabilty between libraries which use a placeholder for values which will become only available in the future, eg. by async operations.

The benefit is, that [`resolve()`](https://github.com/reactphp/promise/pull/78/files#diff-84b928a9905378810b602c8a740b4cbaR13) can properly turn return values from other libraries into react promises. Also, other libraries can use react promises in their way they handle interop promises, eg. through couroutines.

**Comment by WyriHaximus** [comment]
My main concern with not extending the interop promise that we can't communicate both interfaces that way. While all promise classes implement both interfaces, there isn't a single interface that communicates both interfaces. This is mainly coming from my work on @php-api-clients with PHP 7 return types. 

For example the following function has a react promise return type hint as it can be a `FulFilledPromise` or a `RejectedPromise`. (Both of which implement `ReactPromiseInterface` and `AsyncInteropPromise`.)

```php
function r(): ReactPromiseInterface
```

Now the following function only accepts a `AsyncInteropPromise`. 

```php
function g(AsyncInteropPromise $promise): AsyncInteropPromise
```

Technically `g()` should accept the result of `r()` when implementing both but it is imho not clearly communicating it will. When `ReactPromiseInterface` extends `AsyncInteropPromise` `g()` clearly communicates it can and will accept the result of `r()`.

Now this all assumes PHP 7 (which I hope we'll move more towards this year (maybe even PHP 7 only release, but that's a different discussion)) return types and type hints. Having a single type you can hint makes it cleaner for end users IMHO compated to docblocks, or even IDE detection of possible return values. 

--------------

From the work I've been doing over at @php-api-clients perspective depending on how we do the interop matters in user experience.

There are two options, extending would mean I can return `ReactPromiseInterface` which has our rich promise API with all it's functions. While also return `AsyncInteropPromise` through `ReactPromiseInterface` and have interoperability with other possible consumers.

While with implementing both `ReactPromiseInterface` and `AsyncInteropPromise` I would have either return `ReactPromiseInterface` or `AsyncInteropPromise` as return type. Hinting `ReactPromiseInterface` would expose our rich API, while `AsyncInteropPromise` would give interoperability but not our rich API. Now I could hint `AsyncInteropPromise` and wrap everything in a `resolve()`. That IMHO would clutter the examples and but more importantly when applying it on all levels make a unnecessary mess as all `AsyncInteropPromise` returning calls would have to be wrapped into `resolve()`.

--------------

I agree with @jsor that `when()` is for interoperability and we should keep promising `then()` etc as our main API. How every I agree with @clue that this should be documentated very well. 

Just my two euro's.

**Comment by trowski** [comment]
I agree with everything @WyriHaximus mentioned above. React's promises should extend the async-interop interface to make usability easier for everyone.

Linking to the async-interop standard should be sufficient to explain the inclusion of `when()` in React's promise interface. `when()` was designed to be lightweight, primarily for use with coroutines, not as a replacement for `then()` or `done()`. In fact, the name `when()` was chosen to avoid conflicts with the API of any of the existing promise implementations.

**Comment by jsor** [comment]
I disagree for the reason @clue is mentioning: confusion for the consumer.

My proposal here is, that the interop API is completely hidden from the consumer. They should just not use it. It is only intended for other libraries making it easier to consume react's promises through their `resolve()` counterpart.

Regarding @WyriHaximus example:

```php
function g(AsyncInteropPromise $promise): AsyncInteropPromise
```

I think that this is just wrong. Why typehint against the placeholder rather the actual value the function is working on?

IMHO, the correct signature is

```php
function g(MyValue $value): AsyncInteropPromise
```

And the function should be called like

```php
React\Promise\resolve($promise)->done(function (MyValue $myValue) {
    g($myValue);
});

// or in a couroutine based library like amphp

/** @var $myValue MyValue **/
$myValue = yield $promise;
g($myValue);
```

**Comment by trowski** [comment]
The issue with React's promises not extending the async-interop interface is functions declaring `React\Promise\PromiseInterface` as a return type cannot automatically be assumed to have a `when()` method. If this cannot be relied on, we're right back where we started - wrapping one libraries promises with another libraries promises.

**Comment by kelunik** [comment]
> I think that this is just wrong. Why typehint against the placeholder rather the actual value the function is working on?

I think @jsor is absolutely right there. But only as long as it's not some kind of combinator function. Combinator functions also benefit from interoperable promises, alongside with coroutines.

If `ReactPromise` doesn't extend `InteropPromise`, most return type declarations will declare `ReactPromise` I guess instead of `InteropPromise`. In that case, we can't be sure that `ReactPromise` also provides `when()`.

If all of your own implementations implement both interfaces, it doesn't matter to them whether they do that or implement only `ReactPromise` which extends `InteropPromise`.

The only difference is that custom implementations of `ReactPromise` have to provide the interoperable `when()`, too.

For the consumer, nothing changes. All it changes is - as said - custom implementations. If you want to hide the `when()` method, fine, you could add an `@internal` annotation to it to discourage its use. But I honestly thing that this should be solved through documentation explaining what's `when` and what's `then` / `done` / `otherwise`

**Comment by jsor** [comment]
> The issue with React's promises not extending the async-interop interface is functions declaring React\Promise\PromiseInterface as a return type cannot automatically be assumed to have a when() method.

I think this is a good thing because it prevents from mixing up async-interop promises with higher level placeholder abstractions like JS-style promises, coroutines, Rx obervables etc.. The async-interop interface just let you convert between these concepts.

> If this cannot be relied on, we're right back where we started - wrapping one library's promises with another library's promises.

I think that's exactly **not** the purpose of the async-interop interface. Choose the promise/coroutine/rx/whatever-style library you prefer. If you use a dependency which prefers another style, the async-interop interface enables to convert from one style to another.

**Comment by kelunik** [comment]
> I think that's exactly not the purpose of the async-interop interface. Choose the promise/coroutine/rx/whatever-style library you prefer. If you use a dependency which prefers another style, the async-interop interface enables to convert from one style to another.

**That is exactly the aim of the interop spec.** Being able to consume what ever implementation is used. Not adapting everything. If we want to adapt things, we don't need an interop spec, we can just write adapters for the most common libraries.

**Comment by kelunik** [comment]
> Choose the promise/coroutine/rx/whatever-style library you prefer.

Right, that's what the interop spec is for. You can choose whatever you prefer, but you can _always_ consume it via the interface defined in the interop spec. But that's _only_ possible if `ReactPromise` extends `InteropPromise`.

Quoting from the current specification:

> This specification is not designed to replace promise implementations that may be chained. Instead, the common interface may be extended by promise implementations.

**Comment by jsor** [comment]
> Right, that's what the interop spec is for. You can choose whatever you prefer, but you can always consume it via the interface defined in the interop spec.

You could, but why would you? Why would you use `when()` directly and loose all the sugar react/promise, amphp etc. provides?

> But that's only possible if ReactPromise extends InteropPromise.

Why that?

**Comment by bwoebi** [comment]
To be able to pass it directly into generic functionality … Coroutines, combinators etc. They all use the standards sole when() function.

**Comment by kelunik** [comment]
> You could, but why would you? Why would you use when() directly and loose all the sugar react/promise, amphp etc. provides?

Amp doesn't provide any sugar on it's [promise implementation](https://github.com/amphp/amp/blob/master/lib/Internal/Placeholder.php). The sugar are coroutines and combinator functions, they directly consume any promise implementation using `when`.

So the application developer usually doesn't event call `when`. That's also the reason why we don't need any complex chainability.

`react/promise` might want to continue offering its `then` / `done` / `otherwise` API as an alternative to coroutines and thus might extend the `InteropPromise`. Application / library developers might then choose to use that instead of using coroutines.

> Why that?

Because otherwise not every `ReactPromise` is a `InteropPromise` and thus not every `ReactPromise` can be consumed by other libraries such as the Amp combinators.

**Comment by WyriHaximus** [comment]
> > I think that this is just wrong. Why typehint against the placeholder rather the actual value the function is working on?

> I think @jsor is absolutely right there. But only as long as it's not some kind of combinator function. Combinator functions also benefit from interoperable promises, alongside with coroutines.

There is no doubt that @jsor is right. That code snippet was merely to illustrate my point that when `ReactPromiseInterface` extending `InteropPromise` any other interop promise compatible package can consume them. Since `ReactPromiseInterface` is the contract represented by all concrete react promises it is very often used as (return) type hint (either through PHP 7 features or docblocks). 

When `ReactPromiseInterface` extends `InteropPromise` all packages using react's promise package all of the sudden (well after an update) communicate they support interop promises **without any** changes to their code. (That is [146](https://packagist.org/packages/react/promise/dependents) dependents and anything depending on those. (Assuming they hint `ReactPromiseInterface`.))

**Comment by jsor** [comment]
My point is, that there's no such thing as "generic" :)

`then()` etc. **is** our API, `when()` is not and (most probably) won't be. I understand the amphp perspective where you're using a coroutine based style and where it's a lot easier to base the promise implementation on `when()` alone. And i'm perfectly fine with your interpretation of the async-interop interface, but i think we'll just have to agree to disagree here :)

I'm not completely against extending the async-interop interface but this might need some discussion with @reactphp/core.

**Comment by kelunik** [comment]
@jsor We tried repeatedly to get input from you guys, but nothing happened apart from @WyriHaximus participating and the some activity from you on the latest PRs. If you guys do not extend the interface now, then the promise interop project has failed.

> then() etc. is our API, when() is not

Currently, because there's no reason to have `when()` currently. But interoperability brings that reason: interoperability.

> I understand the amphp perspective where you're using a coroutine based style

The goal is enabling that to all promise implementations. So users might continue to use `react/promise`, but additionally use interoperable coroutines.

> and where it's a lot easier to base the promise implementation on when() alone

Nobody suggests to drop `then` and use `when` alone. The suggestion is to support `when` additionally.

**Comment by stefanotorresi** [comment]
As a ReactPHP user, I'd love to see the this proposal implemented.
I hope that @reactphp/core folks realize that ReactPHP is probably the biggest player in async PHP, and the successful adoption of a standard interface stands largely on their shoulder.
I realise you may have some design concerns (which tbh seem minor to me), but please weigh the advantages/disadvantages thoroughly.

File: README.md [hunk 1]
```diff
@@ -512,6 +518,27 @@ The `React\Promise\PromisorInterface` provides a common interface for objects
...
+Note, that although the interface is named *Promise*, it has nothing to do with
+promises commonly known from the
+[CommonJS Promises/A](http://wiki.commonjs.org/wiki/Promises/A),
+[Promises/A+](https://promisesaplus.com) or
+[ECMAScript 2015 (ES6)](http://www.ecma-international.org/ecma-262/6.0/#sec-promise-constructor)
+specifications.
+
+If you're using React/Promise as your primary value placeholder implementation, 
+it is recommended to **not** use the `when()` method directly but the methods
+provided by the [`PromiseInterface`](#promiseinterface).
```
Comment (kelunik) [inline README.md]: I think this can be reworded to: _"If you use `react/promise` implementations, it's recommended to use the high level consumption methods `then()`, `done()`, `otherwise()` and `always()` to consume a promise's value or reason."_

File: README.md [hunk 2]
```diff
@@ -512,6 +518,27 @@ The `React\Promise\PromisorInterface` provides a common interface for objects
...
 
+### AsyncInterop\Promise compatibility
+
+The [`PromiseInterface`](#promiseinterface) extends the
+[`AsyncInterop\Promise`](https://github.com/async-interop/promise) interface and
+thereby all promise implementations from React/Promise implement its `when()`
+method.
+
+The `AsyncInterop\Promise` interface is intended for interoperability with other
+libraries and their combinator and conversion functions.
```
Comment (kelunik) [inline README.md]: Suggestion: _"The `AsyncInterop\Promise` interface allows interoperability with other promise implementations and their combinator and coroutine implementations."_

File: README.md [hunk 3]
```diff
@@ -512,6 +518,27 @@ The `React\Promise\PromisorInterface` provides a common interface for objects
...
+
+Note, that although the interface is named *Promise*, it has nothing to do with
+promises commonly known from the
```
Comment (kelunik) [inline README.md]: I'd reword this paragraph. While the interop promise may have nothing to do with the concrete interface of A+, it's still the same thing. A placeholder representing the result of an async operation.

Comment (kelunik) [inline README.md]: @clue I understand your point, but I'm not sure whether it really helps consumers to understand it. They'll just think _"Why the hell is it called promise if it's something entirely different?!"_ and be more confused than before. We had a really long discussion about the naming in https://github.com/async-interop/promise/issues/4 and another one to switch back to `Promise`, but I can't find that one right now. The reason behind it is that we kept calling it `Promise` while talking about it, because it's just that, just with a different consumer API.

File: README.md (hunk 1 above)
Comment (kelunik) [inline README.md]: Well, doesn't the explicit recommendation against it make you wonder even more why it exists? The alternatives are in no way more _"correct"_ than `when()`, they're just more high-level, and that's how I'd personally word it. You can totally use `when()`, but be aware of it being low-level. I think it makes sense to mention a reason when to use `when()`: It makes sense whenever your function accepts promises instead of direct values, as it can accept any promise then from any vendor.

File: README.md (hunk 3 above)
Comment (clue) [inline README.md]: > *"Why the hell is it called promise if it's something entirely different?!"*

I think this is a very important point here: To me, this whole PR seems to convey the notion that *this* library implements a concept commonly known as "promises", while async-interop/promise *only* offer a low-level API that aims to ease interoperability between different promise vendors.

> We had a really long discussion about the naming in async-interop/promise#4 and another one to switch back to Promise, but I can't find that one right now.

This isn't *exactly transparent* to me either, so I'd very much appreciate if you happen to find this and could link to this here :+1: 

> […] we kept calling it Promise while talking about it, because it's just that, just with a different consumer API.

I understand the reasoning – but would disagree from a consumer perspective. If there's project A (e.g. this project) that describes "promises" and the interop standard that describes a different API then it's not the same from a consumer perspective.

This means that project A should either ditch its description and adopt the interop API or the other way around (which I don't see happening any time soon).

File: README.md (hunk 1 above)
Comment (clue) [inline README.md]: Sorry for my wording, "correct" wasn't intended to mean the alternative is "incorrect", but rather "not recommended and/or endorsed".

> You can totally use when(), but be aware of it being low-level. 

I guess this is the culprit here: Why would react/promise endorse a low-level API to its consumers if it offers a (conceived or otherwise) more powerful and comfortable higher-level API? See also the comments below :+1:

File: README.md (hunk 3 above)
Comment (trowski) [inline README.md]: @clue @kelunik Here is the related PR changing the name back to Promise: https://github.com/async-interop/promise/pull/15

File: README.md (hunk 1 above)
Comment (kelunik) [inline README.md]: It totally depends on what's consumer for you. Is it just application developers or also library developers? The latter will probably want to use `when()` if they also accept other promise implementations instead of wrapping all promises in React's promises. If all you need is a simple callback on completion (no chains), `when()` is pretty equivalent to `always()` / `done()` (depending whether you need the value / reason or not), except that it works on all compatible implementations then, not only React.

In fact, you might want to adapt `done()` to `when()` semantics regarding the error handling to provide a clear separation of concerns. IMO, throwing from `resolve` / `reject` because the consumption of the value failed is wrong. @jsor mentioned that in our face2face meeting as one of the ugliest things in the current promise implementation.

```
<?php

require __DIR__ . "/vendor/autoload.php";

$deferred = new React\Promise\Deferred;

$deferred->promise()->done(function ($value) {
    var_dump("value", $value);
}, function ($reason) {
    throw new Exception;
    var_dump("reason", $reason);
});

try {
    // shouldn't throw here, instead forward to the interop error handler
    $deferred->reject("foobar");
} catch (Exception $e) {
    var_dump("catch", $e);
}
```

Sorry for the lengthy discussion here, should I open a separate issue for that one? It depends on this PR, but I think it could make sense for 3.0.

Comment (jsor) [inline README.md]: No, `done()` !== `when()`. In fact, `done()` *is* the error handler and you must not throw from it. You should only call it from the root of the promise chain. If a rejection ends up here, it's a bug and its the same as if the error handler throws.

```php
$deferred->promise()->done(function ($value) {
    var_dump("value", $value);
}, function ($reason) {
    callEmergency($reason);
});
```

Comment (kelunik) [inline README.md]: @jsor The following examples are pretty much equivalent, no?

Difference to `done()`: The `when()` combines the two callbacks into one error-first callback to make it harder to ignore an error by accident.

```php
$promise->done($onSuccess, $onFailure);
```

```php
$promise->when(function ($reason, $value) use ($onSuccess, $onFailure) {
    $reason ? $onFailure($reason) : $onSuccess($value);
});
```

Difference to `always()`: The `when()` call gets the `$reason` and `$value` passed.

```php
$promise->always($callback);
```

```php
$promise->when($callback);
```

The only difference I can see is the mechanisms when the callbacks throw errors. `when()` requires them to be forwarded to the `ErrorHandler` while `always()` and `done()` let the error bubble up to the producer of the value, so `$deferred->reject()` / `$deferred->resolve()`.

Comment (jsor) [inline README.md]: > The only difference I can see is the mechanisms when the callbacks throw errors. when() requires them to be forwarded to the ErrorHandler while always() and done() let the error bubble up to the producer of the value, so $deferred->reject() / $deferred->resolve().

The difference applies to `when()` and `done()`. `always()` is different as it doesn't let errors bubble up, just like `then()` and `otherwise()`.

Also, throwing from `then()`, `otherwise()` and `always()` is encouraged to propagate a rejection.

Consider the following synchronous code:

```php
try {
    $result = operation();
    $transformedResult = transform($result);
    print_r($transformedResult);
} catch (Throwable $e) {
    handleError($e);
} finally {
    cleanup();
}
```

This is the promise based equivalent:

```php
operation()
    ->then('transform')
    ->then('print_r')
    ->otherwise('handleError') // catch
    ->always('cleanup') // finally
    ->done() // bubble out fatal exceptions thrown from handleError or cleanup
;
```

Comment (jsor) [inline README.md]: Tbh, the two examples are not *exactly* the same as we don't have something like a [try() (or attempt())](http://bluebirdjs.com/docs/api/promise.try.html) function (yet).

```php
React\Promise\attempt('operation')
    ->then('transform')
    ->then('print_r')
    ->otherwise('handleError') // catch
    ->always('cleanup') // finally
    ->done() // bubble out fatal exceptions thrown from handleError or cleanup
;
```

Comment (kelunik) [inline README.md]: @jsor You're right about `always`, thought it would only allow clean ups, missed that it supports forwarding as well.

My main point is only about this statement from your README:

> Calling done() transfers all responsibility for errors to your code. If an error (either a thrown exception or returned rejection) escapes the $onFulfilled or $onRejected callbacks you provide to done, it will be rethrown in an uncatchable way causing a fatal error.

Because in fact, those errors aren't uncatchable, they bubble up to `$deferred->resolve()` and then further up from there. To prevent exactly the bubbling up to `$deferred->resolve()`, the interop spec introduced the `ErrorHandler` which can either log such an error gracefully or abort the script if not such handler is defined.

Comment (jsor) [inline README.md]: > Because in fact, those errors aren't uncatchable, they bubble up to $deferred->resolve() and then further up from there.

Yes, they are catchable if resolution happens synchronously. *catchable* is a bit misleading here as it means *uncatchable* by any handler, not (only) from try/catch.

> the interop spec introduced the ErrorHandler which can either log such an error gracefully or abort the script if not such handler is defined.

Nothing prevents you from transferring the error to the ErrorHandler from inside done().

Comment (kelunik) [inline README.md]: This has really become quite lengthy now. [Want to continue in chat?](http://chat.stackoverflow.com/rooms/11/php)

Comment (jsor) [inline README.md]: I might hop in when time allows. But, we're not going to change the behavior of `done()`. It's a defacto standard among promise libraries and it's is in line with all implementations i'm aware of.

Comment (kelunik) [inline README.md]: @jsor Bluebird for one doesn't let errors from `done` bubble up to the `resolve` call, it executes [`async.fatalError`](https://github.com/petkaantonov/bluebird/blob/master/src/async.js#L49-L57) and thus exists. If it's not in node, it doesn't let it bubble up either, it defers it to the next tick and throws from a `0` timeout callback, similar what we had before we had the error handler. But as we don't have the a loop in the language, we wanted to decouple those and introduced a separate error handler.

**Comment by bwoebi** [comment]
@clue What's the current progress? Is this PR ready to be merged?
If you do, we also could look forward to officially release the 1.0.0 tag for AsyncInterop\Promise.

**Comment by kelunik** [comment]
> I guess this is the culprit here: Why would react/promise endorse a low-level API to its consumers if it offers a (conceived or otherwise) more powerful and comfortable higher-level API?

Because it makes React compatible with coroutine implementations and generic combinator functions.

**Comment by trowski** [comment]
@clue By implementing `AsyncInterop\Promise`, react/promise is not endorsing the low-level API to it's consumers, rather acknowledging that a React promise may also be used anywhere that an interop promise can be used. These places will include, but not be limited to, coroutines, combinators, and (with support from Rx) observables.

When an API returns a React promise, a user should be encouraged to use the high-level API of the React promise. However, that same user can also know that they may use that promise in conjunction with any other library that accepts interop promises.

`AsyncInterop\Promise` only contains a low-level API because that is all that is necessary for interop. Any other promise methods can be implemented from the `when()` method. As a purely interop interface, we did not want to force a particular interface on promise libraries. What if the interop interface included methods that conflicted with those of React? We purposely avoided this problem with a low-level API consisting of a single, uniquely named method. Now this seems to be a blocker for adoption… to me this makes little sense.

Documentation of `when()` has been specifically excluded from this library, with which I take no issue. Use of `when()` should generally be discouraged by users. It exists for libraries, not applications. Amp discourages this by encouraging users to use promises in coroutines. React discourages this through their high-level `then()`/`done()` API. Right now it is difficult to intermix code from these libraries. With this PR, suddenly this will become trivial. `React\Promise\resolve()` could be used to transform an interop promise returned from Amp and use React's API. Conversely, a user of Amp could use a React promise directly in a coroutine.

Hopefully this will help you understand the purpose of this PR and the async-interop group. It is not to replace the API of any particular group, but rather allow sharing of code between libraries of different groups. We want to push async in PHP forward, making it as easy as possible to share code and collaborate, allowing each group to code to their own style and preferences, without having to worry about compatibility with a particular event loop or promise implementation.

**Comment by jsor** [comment]
Closing this for now. It might be reconsidered at a later point.
